### PR TITLE
[MM-29362] Allow filtering by hosting type

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
This PR adds a `hosting` field to the stored plugins. If it's set to either `cloud` or `on-prem` the plugin will be limited to that kind of installation.

The PR does not contain any changes to the current list of plugins. These will be added in a separate PR.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-29362

#### Related Pull Requests

https://github.com/mattermost/mattermost-server/pull/15913

### CODE REVIEW

File: cmd/generator/add.go [hunk 1]
```diff
@@ -52,6 +54,20 @@ var addCmd = &cobra.Command{
...
+			return err
+		}
+
+		onPrem, err := command.Flags().GetBool("on-prem")
+		if err != nil {
+			return err
+		}
+
+		if cloud && onPrem {
+			return errors.New("if you want to make a plugin availed for cloud and on-prem, just drop both flags")
```
Comment (catalintomai) [inline cmd/generator/add.go:68]:
```suggestion
			return errors.New("if you want to make a plugin available for cloud and on-prem, just drop both flags")
```

Comment (hanzei) [reply cmd/generator/add.go]: Good catch :+1:

**Review by catalintomai — APPROVED** [review]

File: internal/api/plugins_test.go [hunk 2]
```diff
@@ -395,6 +415,54 @@ func TestPlugins(t *testing.T) {
...
+
+			plugins, err := client.GetPlugins(&api.GetPluginsRequest{
+				PerPage: -1,
+				Cloud:   true,
+			})
+			require.NoError(t, err)
+			require.ElementsMatch(t, []*model.Plugin{plugin1V3Min515, plugin2V1Min516, plugin3V3Min517}, plugins)
+		})
+
+		t.Run("on-prem plugin is not return for on-prem instance", func(t *testing.T) {
```
Comment (mickmister) [inline internal/api/plugins_test.go:454]:
```suggestion
		t.Run("on-prem only plugin is not return for cloud instance", func(t *testing.T) {
```

File: internal/model/plugin_test.go [hunk 3]
```diff
@@ -156,6 +157,7 @@ func TestPluginsToWriter(t *testing.T) {
       "version": "2.0.0"
     },
     "enterprise": false,
+    "hosting": "",
```
Comment (mickmister) [inline internal/model/plugin_test.go:160]: Does it make sense to have one test assertion here for the hosting being cloud or on-prem? As opposed to only having the empty string case.

**Review by mickmister — APPROVED** [review]
LGTM, just some non-blocking suggestions on tests

File: internal/model/plugin_test.go (hunk 3 above)
Comment (hanzei) [reply internal/model/plugin_test.go]: Makes sense. Added :+1:

File: internal/api/plugins_test.go (hunk 2 above)
Comment (hanzei) [reply internal/api/plugins_test.go]: Good catch :+1:

**Comment by hanzei** [comment]
@DHaussermann I've added test steps to this in https://mattermost.atlassian.net/browse/MM-29362

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
This is working as expected when I followed the test steps provided in the Jira task linked. We can now set where a plugin is `cloud` or `on-prem` in the marketplace config. 
LGTM!

Thanks @hanzei for steps and guidance on this one.

**Pull request merged**